File: src/models/sync.ts
import { Transaction } from "./transaction";

export interface SyncSession {
  pin: string;
  token?: string | null;
  createdAt: number;
  expiresAt: number;
  isActive: boolean;
  deviceName?: string | null;
}

export interface SyncAuthRequest {
  pin: string;
  deviceName: string;
}

export interface SyncAuthResponse {
  success: boolean;
  token?: string | null;
  expiresIn?: number | null; // seconds
  message: string;
}

export interface SyncDataPayload {
  transactions: Transaction[];
  deviceName: string;
  timestamp: number;
}

export interface SyncDataResponse {
  success: boolean;
  pendingApproval: boolean;
  conflicts: SyncConflict[];
  message: string;
}

export interface SyncConflict {
  conflictType: ConflictType;
  transactionId: string;
  description: string;
  suggestion?: string | null;
}

export type ConflictType =
  | {
      type: "insufficientBalance";
      accountId: string;
      accountName: string;
      currentBalance: number;
      required: number;
    }
  | {
      type: "unknownItem";
      itemName: string;
      suggestedMatches: ItemMatch[];
    }
  | {
      type: "duplicateTransaction";
    }
  | {
      type: "invalidAccount";
    };

export interface ItemMatch {
  itemId: string;
  name: string;
  brand?: string | null;
  similarityScore: number; // 0.0 - 1.0
}

export interface PendingSyncData {
  id: string;
  payload: SyncDataPayload;
  conflicts: SyncConflict[];
  receivedAt: number;
  deviceName: string;
}

export interface SyncApprovalRequest {
  syncId: string;
  approved: boolean;
  conflictResolutions: Record<string, ConflictResolution>;
}

export type ConflictResolution =
  | { type: "skipTransaction" }
  | { type: "adjustAmount"; newAmount: number }
  | { type: "mapItem"; itemId: string }
  | { type: "createNewItem" };

File: src-tauri/src/models/sync.rs
// src/models/sync.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use crate::models::transaction::Transaction;

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub struct SyncSession {
    pub pin: String,
    pub token: Option<String>,
    pub created_at: u64,
    pub expires_at: u64,
    pub is_active: bool,
    pub device_name: Option<String>,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub struct SyncAuthRequest {
    pub pin: String,
    pub device_name: String,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub struct SyncAuthResponse {
    pub success: bool,
    pub token: Option<String>,
    pub expires_in: Option<u64>, // segundos
    pub message: String,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub struct SyncDataPayload {
    pub transactions: Vec<Transaction>,
    pub device_name: String,
    pub timestamp: u64,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub struct SyncDataResponse {
    pub success: bool,
    pub pending_approval: bool,
    pub conflicts: Vec<SyncConflict>,
    pub message: String,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub struct SyncConflict {
    pub conflict_type: ConflictType,
    pub transaction_id: String,
    pub description: String,
    pub suggestion: Option<String>,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub enum ConflictType {
    InsufficientBalance {
        account_id: String,
        account_name: String,
        current_balance: f64,
        required: f64,
    },
    UnknownItem {
        item_name: String,
        suggested_matches: Vec<ItemMatch>,
    },
    DuplicateTransaction,
    InvalidAccount,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub struct ItemMatch {
    pub item_id: String,
    pub name: String,
    pub brand: Option<String>,
    pub similarity_score: f32, // 0.0 - 1.0
}

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub struct PendingSyncData {
    pub id: String,
    pub payload: SyncDataPayload,
    pub conflicts: Vec<SyncConflict>,
    pub received_at: u64,
    pub device_name: String,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub struct SyncApprovalRequest {
    pub sync_id: String,
    pub approved: bool,
    pub conflict_resolutions: HashMap<String, ConflictResolution>,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub enum ConflictResolution {
    SkipTransaction,
    AdjustAmount { new_amount: f64 },
    MapItem { item_id: String },
    CreateNewItem,
}
